Ionyshch/install

ionyshch-prismsl opened the pull request:
EDGE-111 Create build scripts for building Connect SDK with Hanwha WN toolchain

This is work in progress: files and folders are moved, but they have to be checked and fixed, if necessary. Working part is hanwha-wn platform. To test, one has to install OpenSDK according to installation guide (accompanying OpenSDK)

Need feedback sooner rather than later.

ionyshch-prismsl committed:
Build boost

ionyshch-prismsl committed:
EDGE-111 Create build scripts for building Connect SDK with Hanwha WN toolchain

ionyshch-prismsl committed:
Moved files are deleted

ionyshch-prismsl committed:
Added moved tests and examples. Move ios specific shell script

dbus-prsm commented on CMakeLists.txt:
```diff
@@ -2,84 +2,66 @@
+    if (NOT DEFINED PRISM_PLATFORM)
+        message (FATAL_ERROR "When crosscompiling, PRISM_PLATFORM must be set")
+    endif ()  
+elseif (NOT DEFINED PRISM_PLATFORM)
```
Typo.
`elseif (CMAKE_CROSSCOMPILING)`
should be here

ionyshch-prismsl replied:
it's elseif (), not else()

dbus-prsm commented on platforms/macos/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+findBoostCurl()
+buildSdk()
+
+add_subdirectory(../shared/tests)
```
I'm getting cmake error for these lines:
`add_subdirectory not given a binary directory but the given source
  directory "./connect/platforms/shared/tests" is not a
  subdirectory of "./connect/platforms/macos".  When
  specifying an out-of-tree source a binary directory must be explicitly
  specified.`
cmake version:  3.7.1

ionyshch-prismsl replied:
Will fix today. Sharing examples/tests code looks somewhat weird from cmake's point of view.

dbus-prsm commented on platforms/hanwha-wn/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+        --without-mpi
+        --disable-icu
+        --prefix=${CMAKE_CURRENT_BINARY_DIR}/install
+        -j8
```
Hardcode. You can use ProcessorCount() function

ionyshch-prismsl replied:
OK

dbus-prsm commented on platforms/hanwha-wn/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+file(WRITE ${BOOST_BUILD_PATH}/user-config.jam "using gcc : arm : \"$ENV{WN}/arm-linux-gnueabi-g++\"")
+
+ExternalProject_Add(boost
+    URL https://sourceforge.net/projects/boost/files/boost/1.63.0/boost_1_63_0.tar.bz2
```
Is it possible to use the same Boost version for SDK and hanwha?

ionyshch-prismsl replied:
For now we have minimum requirement for Boost in find_package(): it's 1.60.0
We may use whatever version suits us, I just used the latest one.

dbus-prsm commented on platforms/hanwha-wn/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+    UPDATE_COMMAND ""
+    PATCH_COMMAND ""
+    CONFIGURE_COMMAND ./bootstrap.sh
+    BUILD_COMMAND ./b2 install
```
Do we really need to build all of the rest Boost libs here?

Yurmusp replied:
How much time it takes to perform this build? Do we really want it downloads and builds from that URL? We can be considered DDOS attackers if we will be downloading it 10 times a day or more.
I guess it shall be in our Artifactory or in our GDrive.

ionyshch-prismsl replied:
I will make it to build only chrono for now. 
I guess downloading from GDrive is OK, provided it doesn't require authentication.

Yurmusp replied:
You can do public files in Prism GDrive. So, it shall work.

ionyshch-prismsl replied:
It couldn't download it with wget. The reason presumably, that drive shows warning about file being too large to check for viruses. Have to research more about it.

Yurmusp replied:
I looked what people do with it and there is no solution for GDrive it seems. People say it can work with DropBox. We have DropBox, so lets try it.

ionyshch-prismsl replied:
I'll try.
I'd like some feedback on having "download cache". Is ~/Downloads a suitable place for it? Or should we move it into root of Connect SDK as we are doing for install folder?

dbus-prsm replied:
I think it's a good idea to keep everything in the same directory. I wouldn't expect from build script to put something into my ~/Downloads directory.
But for automatic server build, it would be useful to share third party artifacts between different builds.

Yurmusp replied:
I think for today just some temp folder relative to build root shall be fine. Eventually, we will move to Ivy (or something similar) with artifactory that will maintain such cache implicitly by itself.

dbus-prsm commented on platforms/hanwha-wn/toolchain.cmake:
```diff
@@ -0,0 +1,21 @@
+# this one is important
+SET(CMAKE_SYSTEM_NAME Linux)
+#this one not so much
+SET(CMAKE_SYSTEM_VERSION 1)
```
It's not clear, why did you override OS version here?

ionyshch-prismsl replied:
It's from toolchain.cmake example in CMake online docs.

Yurmusp commented on CMakeLists.txt:
```diff
@@ -2,84 +2,66 @@
+    find_program (CMAKE_UNAME uname /bin /usr/bin /usr/local/bin)
+    if (CMAKE_UNAME)
+	    exec_program(uname OUTPUT_VARIABLE PRISM_HOST_OS_NAME)
+	    if (PRISM_HOST_OS_NAME STREQUAL "Darwin")
```
This platform config code shall be somewhere in platform/shared
Not in root CMakeLists.txt

ionyshch-prismsl replied:
We only use this piece of code here for "dispatching" into particular platform.
Shared code is the one used by two or more platforms. Function buildSdk() is an example

Yurmusp replied:
Exactly. Platforms dispatching code shall be in some file in platforms/
We shall not litter root CMakelists.txt with it.

ionyshch-prismsl replied:
I don't understand the point. Root file **is** for dispatching. And suggestion is to move the code, that isn't used anywhere else. What are benefits? The drawback is that one has now open two files (or three) to understand, what's going on. Let's keep logics, where it belongs, keep it readable and don't split two dozens of lines across multiple files

ionyshch-prismsl replied:
I put in platforms/shared code and build files, possibly including functions and macro, that are used in two or more, but not all platform specific build files. Non reusable code and code, that is applicable to all platforms e.g. buildSdk() should IMO be in root cmakelists. Thus we use minimum number of files and keep it readable.

Yurmusp replied:
It all depends how much stuff is in root file. If we have too much of it, I would prefer to move this kind of logic into platform/ . Even if it is one more file, reference and so on.
Logically, platform dispatching is platform related and it belongs to platform/, not to root file.
Root file is a connecting point for everything and it shall have minimum amount of logic.

ionyshch-prismsl replied:
For now root file is small. Dispatching part is made in a way, that adding new platforms doesn't require modifying it. 
Let's keep it as-is for now and if root file grows, consider moving parts to other file(s).

Yurmusp replied:
If you remember in cpp-engine & PVB, build scripts where shared and were in SVN linked project similar to git submodule. We may need same thing here soon. In this case all this platform selection code shall be there. Because otherwise we will have code duplication. Until root file is small and platform/ is not submodule we can live like it is, but it is not clean solution I think.

ionyshch-prismsl committed:
Fixes according to comments.
Boost source archive is looked for in ~/Downloads, if not found, attempt to download.
Added install instructions. Default install location ~/prism/<platform>/

dbus-prsm commented on CMakeLists.txt:
```diff
@@ -2,84 +2,34 @@
+if (DEFINED ENV{PRISM_INSTALL_DIR})
+    set (CMAKE_INSTALL_PREFIX $ENV{PRISM_INSTALL_DIR}/${PRISM_PLATFORM})
 else ()
+    set (CMAKE_INSTALL_PREFIX $ENV{HOME}/prism/${PRISM_PLATFORM})
```
I think we should install inside our source tree by default, not into user's home.

Yurmusp replied:
I also think that Install location shall be relative to root folder of the build. Not relative to user home dir.

ionyshch-prismsl replied:
Ok

dbus-prsm commented on platforms/hanwha-wn/CMakeLists.txt:
```diff
@@ -0,0 +1,51 @@
+
+file(WRITE ${BOOST_BUILD_PATH}/user-config.jam "using gcc : arm : \"$ENV{WN}/arm-linux-gnueabi-g++\"")
+
+set (BOOST_ARCH_NAME boost_1_63_0.tar.bz2)
```
Maybe we should try Boost libs we're using for SDK build. So we'll not have to deal with Boost downloading/build here.

Yurmusp replied:
Not sure how Ivan, but I did not understand Denis comment.

ionyshch-prismsl replied:
These build files are for building Connect SDK. Code and build files for hanwha integration app will be in a separate private repo, and will use boost libraries built here

ionyshch-prismsl committed:
Boost source is now downloaded from DropBox
Install folder is in the root of Connect SDK folder.
Downloads (cache) folder in the root of Connect SDK folder.
Root CMakeLists.txt for now contains code that is used by all platforms.
We may decide to move parts of it to other file(s), if code amount increases.

ionyshch-prismsl committed:
Tested building for iOS. Readme file updated.
Added copyright to our files.
Added license and download info for 3rd party script

ionyshch-prismsl committed:
Fixed formating in .md file
